File: leetcode/tx50/2-两数相加/2.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:

        extra = 0
        head = ListNode(0)
        node = head

        while l1 != None or l2 != None or extra:

            val1 = l1.val if l1 else 0
            val2 = l2.val if l2 else 0
            l1   = l1.next if l1 else None
            l2   = l2.next if l2 else None

            node.val = (val1 + val2 + extra) % 10
            extra    = (val1 + val2 + extra) // 10

            if l1 == None and l2 == None and extra == 0:
                node.next = None
                break

            tmp  = ListNode(0)
            node.next = tmp
            node = tmp

        return head

File: leetcode/tx50/2-两数相加/2.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {

        int extra = 0;
        int val1  = 0;
        int val2  = 0;

        ListNode head = new ListNode(0);
        ListNode node = head;

        while(l1 != null || l2 != null || extra != 0){

            if(l1 != null){
                val1 = l1.val;
                l1   = l1.next;
            }else{
                val1 = 0;
            }
            if(l2 != null){
                val2 = l2.val;
                l2   = l2.next;
            }else{
                val2 = 0;
            }

            node.val  = (val1 + val2 + extra) % 10;
            extra    = (val1 + val2 + extra) / 10;

            if(l1 == null && l2 == null && extra == 0){
                node.next = null;
                break;
            }

            ListNode tmp = new ListNode(0);
            node.next = tmp;
            node = tmp;

        }
        return head;
    }
}
